applications/minotari_node/src/stratum/tari_sha3x_stratum_handler.rs

```diff
@@ -0,0 +1,356 @@
+            extra_nonce_hex: job_record.xn.to_string(),
+            height: job_record.height,
+        })
+    }
+
+    async fn check_notify_needed(&self, last_job_id: String) -> Result<Option<NotifyResponse>, anyhow::Error> {
+        let last_job = self.job_repository.get_job(last_job_id.clone()).await?;
+
+        if last_job.is_none() {
+            return Err(anyhow::anyhow!("Job with id {} not found", last_job_id));
+        }
+
+        let last_job = last_job.unwrap();
+
+        let (best_height, best_block_header) = self.block_template_repository.get_tip().await?;
+
+        if last_job.height == best_height.saturating_add(1) && last_job.prev_block_hash == best_block_header {
+            // No need to notify, the job is still valid
+            return Ok(None);
+        }
+
+        // Else, let's make a new job.
+        let job_record = create_job_from_blob(
+            &self.block_template_repository,
+            last_job.pow_algo,
+            last_job.miner_address.clone(),
+            Some(last_job.xn.clone()),
+        )
+        .await?;
+        self.job_repository.insert_job(job_record.clone()).await?;
+        Ok(Some(NotifyResponse {
+            job_id: job_record.job_id,
+            height: job_record.height,
+            blob: hex::encode(job_record.blob),
+            extra_nonce_hex: hex::encode(job_record.xn.to_le_bytes()),
+        }))
+    }
+}
+
+async fn create_job_from_blob<TBlockRepo: BlockTemplateRepository>(
+    block_template_repository: &TBlockRepo,
+    algo: PowAlgorithm,
+    address: TariAddress,
+    _nonce: Option<u16>,
+) -> anyhow::Result<Job> {
+    let BlockTemplate {
+        blob,
+        height,
+        target,
+        prev_block_hash,
+        ..
+    } = block_template_repository
+        .get_block_template(algo, Some(address.clone()))
+        .await?;
+    let mut r = OsRng;
+    let job_id = hex::encode(r.next_u64().to_le_bytes());
+
+    let id = hex::encode(r.next_u64().to_le_bytes());
+    let random_bytes = rand::random::<u16>();
+    let job_target = hex::encode((u64::MAX / target).to_le_bytes());
+    let chain_target = hex::encode(target.to_le_bytes());
+
+    if blob.is_empty() {
+        return Err(anyhow::anyhow!("No blob available for the latest block"));
+    }
+    let original_mining_hash = blob.clone();
+    let blob = if algo == PowAlgorithm::RandomXT {
+        // The format is:
+        // | 1 byte | 1 byte | 1 bytes | 32 bytes | 8 bytes | 1 byte | 32 bytes|
+        // | major version | minor version | timestamp | mining_hash | nonce (big endian) | pow_algo | pow_data, excluding algo, padded to 32 bytes |
+        //
+        // Major version: 0
+        // Minor version: 0
+        // Timestamp: 0
+        let mut final_blob = vec![0u8; 76];
+        final_blob[0] = 0; // Major version
+        final_blob[1] = 0; // Minor version
+        final_blob[2] = 0; // Timestamp
+        final_blob[3..35].copy_from_slice(&blob); // Mining hash
```

Comment:
We should verify `blob` length to be exactly 32 bytes
```suggestion
        final_blob[3..35].copy_from_slice(&blob); // Mining hash
```